src/deck/core/deckctrl_dfu_memory.c

```diff
@@ -0,0 +1,190 @@
+#define CMD_ENTER_FIRMWARE_MODE (1 << 1)
+
+static uint8_t nbrOfDeckCtrl(void) {
+    uint8_t deckCtrlCount = 0;
+    for (int i = 0; i < deckCount(); i++) {
+        DeckInfo *deck = deckInfo(i);
+        if (strcmp(deck->discoveryBackend->name, "deckctrl") == 0) {
+            deckCtrlCount++;
+        }
+    }
+
+    return deckCtrlCount;
+}
+
+static uint8_t getStatus(void) {
+    uint8_t status = 0;
+
+    if (nbrOfDeckCtrl() <= 1) {
+        status = status | STATUS_CAN_ENABLE_DFU;
+    } else {
+        status = status & (uint8_t)~STATUS_CAN_ENABLE_DFU;
+    }
+
+    uint8_t dummy;
+    if (dfu_i2c_read(DFU_STM32C0_I2C_ADDRESS, 0, &dummy, 1)) {
+        status = status | STATUS_IN_DFU_MODE;
+        status = status & (uint8_t)~STATUS_CAN_ENABLE_DFU;
+    } else {
+        status = status & (uint8_t)~STATUS_IN_DFU_MODE;
+    }
+
+    return status;
+}
+
+// Calling this function will cut the power to VCC including this MCU and the
+// DeckCtrl, causing a reset. The DeckCtrl will then enter DFU mode on next
+// power-up.
+static void enableDFUViaNRF(void * arg) {
+    bool enable = (bool) arg;
+    vTaskDelay(pdMS_TO_TICKS(10)); // Give some time for the ongoing write to complete
+    SyslinkPacket slp;
+    slp.type = SYSLINK_PM_DECKCTRL_DFU;
+    slp.length = 1;
+    slp.data[0] = enable?0x01:0x00; // 1 to enter DFU, 0 to enter firmware mode
+    syslinkSendPacket(&slp);
+}
+
+bool handleMemRead(const uint32_t memAddr, const uint8_t readLen, uint8_t* buffer) {
+
+    if (memAddr < DECK_CTRL_MEM_OFFSET) {
+        for (unsigned int i = 0; i < readLen; i++) {
+            if (memAddr + i == VERSION_MEMORY_BYTE) {
+                buffer[i] = VERSION;
+            } else if (memAddr + i == DECKCTRL_MEMORY_BYTE) {
+                buffer[i] = nbrOfDeckCtrl();
+            } else if (memAddr + i == STATUS_MEMORY_BYTE) {
+                buffer[i] = getStatus();
+            } else {
+                buffer[i] = 0x00;
+            }
+        }
+    } else if (memAddr >= DECK_CTRL_MEM_OFFSET &&
+               memAddr < (DECK_CTRL_MEM_OFFSET + DECK_CTRL_MEM_SIZE)) {
+
+        uint32_t dfuMemAddr = memAddr - DECK_CTRL_MEM_OFFSET;
+        bool result = dfu_i2c_read(DFU_STM32C0_I2C_ADDRESS, dfuMemAddr, buffer, readLen);
+        if (!result) {
+            return false;
+        }
+    } else {
+        return false;
+    }
+    
+    return true;
+}
+
+bool handleMemWrite(const uint32_t memAddr, const uint8_t writeLen, const uint8_t* buffer) {
+
+    if (memAddr < DECK_CTRL_MEM_OFFSET) {
+        for (unsigned int i = 0; i < writeLen; i++) {
+            if (memAddr + i == CMD_MEMORY_BYTE) {
+                if (
+                    (buffer[i] & CMD_ENTER_FIRMWARE_MODE) != 0 &&
+                    (buffer[i] & CMD_ENTER_DFU_MODE) != 0)
+                {
+                    return false;
+                } else if ((buffer[i] & CMD_ENTER_FIRMWARE_MODE) != 0) {
+                    bool shouldEnable = false;
```

Comment:
Is there a chance this variable will be overwritten on the stack before it is used?